download/src/lib.rs

```diff
@@ -47,12 +47,13 @@ fn download_with_backend(
     }
 }
 
+type EventResult<'a> = &'a dyn Fn(Event<'_>) -> Result<()>;
```

Comment:
We had some discussions on this change before: https://github.com/rust-lang/rustup/pull/3096#discussion_r1005434248
```suggestion
type DownloadCallback<'a> = &'a dyn Fn(Event<'_>) -> Result<()>;
```